### PR TITLE
Some React JSX Rule

### PR DESCRIPTION
**Pull request by vandyw95**

### CODE REVIEW

File: rules/react.js [hunk 1]
```diff
@@ -104,5 +104,195 @@ module.exports = {
...
+    // 'react/jsx-max-props-per-line': [
+    //   'error',
+    //   {
+    //     maximum: 1,
+    //     when: 'multiline',
+    //   },
+    // ],
+    //
+    // Set to "always" so that new props will always on new line if there are more than 1 prop
+    'react/jsx-max-props-per-line': [
```
Comment (tavurth) [inline rules/react.js]: Did you manage to fix this so it works with prettier?

**Review by tavurth — DISMISSED** [review]

File: rules/react.js (hunk 1 above)
Comment (vandyw95) [reply rules/react.js]: Oops.. missed that documentation in eslint-config-prettier.
Moved to "TODO" section to be enabled when we moved into prettier-eslint to do the formatting.

Comment (vandyw95) [reply rules/react.js]: @tavurth

File: rules/react.js [hunk 2]
```diff
@@ -104,5 +104,195 @@ module.exports = {
     // Some legacy code heavy use indexes for keys. It is complicated to convert to something else
     // expecially when data is client side derived.
     'react/no-array-index-key': 'warn',
```
Comment (tavurth) [inline rules/react.js:106]: So this will give us a warning right? Index bugs are really awful to debug and I want at least to see a warning.

File: rules/react.js [hunk 3]
```diff
@@ -104,5 +104,195 @@ module.exports = {
...
+    //   },
+    // ],
+
+    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-direct-mutation-state.md
+    // AirBnB
+    // 'react/no-direct-mutation-state': 'off',
+    // ============
+    // Proposed:
+    // Enforce immutability
+    'react/no-direct-mutation-state': 'error',
```
Comment (tavurth) [inline rules/react.js:165]: Great!

File: rules/react.js [hunk 4]
```diff
@@ -104,5 +104,195 @@ module.exports = {
...
+    // 'react/jsx-no-literals': ['off', { noStrings: true }],
+    // ============
+    // Proposed:
+    // Always wrap sting inside literal
+    // GOOD:
+    // <span>{'Hello'}</span>
+    // BAD:
+    // <span>Hello</span>
+    // WHY: consistency and make it more specific to JSX's literal rather than string literal
+    'react/jsx-no-literals': ['error', { noStrings: true }],
```
Comment (tavurth) [inline rules/react.js:209]: Should be groomed

File: rules/react.js [hunk 5]
```diff
@@ -104,5 +104,195 @@ module.exports = {
...
+
+    // ==============================================================================
+    // ==============================================================================
+    // TODO: Enable when moved to prettier-eslint package instead of current config
```
Comment (tavurth) [inline rules/react.js]: Can we move these comments out and into another PR.
We should not commit commented code

**Review by tavurth — DISMISSED** [review]

File: rules/react.js (hunk 4 above)
Comment (tavurth) [reply rules/react.js:209]: @JustBeard

File: rules/react.js (hunk 2 above)
Comment (vandyw95) [reply rules/react.js:106]: I think so

File: rules/react.js (hunk 5 above)
Comment (vandyw95) [reply rules/react.js]: I will separate it into TODO.js

Comment (tavurth) [reply rules/react.js]: Actually I think we should make a separate PR. Branches are for code, not comments

**Comment by tavurth** [comment]
Looks much better now @vandyw95 Thank you!

**Review by tavurth — APPROVED** [review]

**Comment by tavurth** [comment]
I'd still like to look into the jsx-no-string-literals

**Review by Jedmed — COMMENTED** [review]
I think we should keep jsx-no-literals off. I think it’s adding unnecessary complexity to strings in jsx.

File: rules/react.js [hunk 6]
```diff
@@ -104,5 +104,108 @@ module.exports = {
...
+    // Enforce styling inside styled-component rather than class
+    // We can pass custom props to styled-component and assign logic there
+    // Span allowed for our icon set like linear icons but we can also specify the required icon
+    // inside styled-component attrs but that means we need to create a lot of icon component.
+    // If no additional styling required, using className for span icon probably good enough.
+    // Also forbid using "style" as component's props.
+    'react/forbid-component-props': [
+      'error',
+      {
+        forbid: [{ propName: 'className', allowedFor: ['span'] }, { propName: 'style' }],
```
Comment (beermindnor) [inline rules/react.js:182]: Linear Icon could attach by props by styled-component
```
const LargeCircle = styled.span.attrs({ className: 'lnr lnr-circle' })'
  font-size: 36px;
  color: #ff0000;
`
```
Another case that when using third-party className is seems to be best way to control their styles. For `react-transition-group` also className needed they using className as a transition prefix too.
So I think className should be very generic already with JSX and should not strict on it. I'm open for groom this one too.

**Comment by JustBeard** [comment]
https://github.com/EkoCommunications/EkoESLintConfig/pull/6
